open-machine-learning-jupyter-book/deep-learning/CNN.md

```diff
@@ -13,4 +13,1201 @@ kernelspec:
+
+-- Andrew Ng
+```
+
+
+
+Convolutional Neural Networks (CNNs) are responsible for the latest major breakthroughs in image recognition in the past few years.
+
+In mathematics, a convolution is a function which is applied over the output of another function. In our case, we will consider applying a matrix mutliplication (filter) across an image. See the below diagram for an example of how this may work.
+
+
+
+:::{figure-md} 01_intro_cnn-dl
+<img src="../../images/deep-learning/CNN/01_intro_cnn.png" width="90%" class="bg-white mb-1">
+
+Illustration of matrix mutliplication (filter) in CNN {cite}`reluwiki`
+:::
+
+
+
+<iframe src="../html-iframes/conv-demo/index.html" width="105%" height="700px;" style="border:none;"></iframe>
+
+
+CNNs generally follow a structure. The main convolutional setup is (input array) -> (convolutional filter layer) -> (Pooling) -> (Activation layer). The above diagram depicts how a convolutional layer may create one feature. Generally, filters are multidimensional and end up creating many features. It is also common to have a completely separate filter-feature creator of different sizes acting on the same layer. After this convolutional filter, it is common to apply a pooling layer. This pooling may be a max-pooling or an average pooling or another aggregation. One of the key concepts here is that the pooling layer has no parameters while decreasing the layer size. See the below diagram for an example of max-pooling.
+
+
+
+:::{figure-md} 01_intro_cnn2-dl
+<img src="../../images/deep-learning/CNN/01_intro_cnn2.png" width="90%" class="bg-white mb-1">
+
+Illustration of max pooling
+:::
+
+
+After the max pooling, there is generally an activation layer. One of the more common activation layers is the ReLU (Rectified Linear Unit) {cite}`reluwiki` .
+
+## MNIST handwritten digits
+Here we illustrate how to use a simple CNN with three convolutional units to predict the MNIST handwritten digits. 
+
+
+```{note}
+There is good reason why this dataset is used like the 'hello world' of image recognition, it is fairly compact while having a decent amount of training, test, and validation data. It only has one channel (black and white) and only ten possible outputs (0-9).
+```
+
+When the script is done training the model, you should see similar output to the following graphs.
+
+
+
+:::{figure-md} 02_cnn1_loss_acc-dl
+<img src="../../images/deep-learning/CNN/02_cnn1_loss_acc.png" width="90%" class="bg-white mb-1">
+
+Train MINIST dataset with CNN: accuracy and loss
+:::
+
+
+Training and test loss (left) and test batch accuracy (right).
+
+
+
+:::{figure-md} 02_cnn1_mnist_output-dl
+<img src="../../images/deep-learning/CNN/02_cnn1_mnist_output.png" width="90%" class="bg-white mb-1">
+
+Train MINIST dataset with CNN: prediction output
+:::
+
+
+A random set of 6 digits with actual and predicted labels. You can see a prediction failure in the lower right box.
+
+### Code

```

Comment:
Can we break down this big code section to multiple {code-cell}, and add some explanation correspondingly? It is very difficult to follow such a long code in a book style doc.